### PR TITLE
Optimize Actor::ConstructMeshes

### PR DESCRIPTION
**Pull request by amzn-tommy**
There are two optimizations in the PR
1. To count the number of unique jointId's, ConstructMeshes was previously creating a set of jointId's and then checking the size of that set. Instead, CalcLocalJointIndexCountAndHighestIndex will create a bitset large enough to hold one bit for each possible values of uint16, set the bit for each id that is used, and then count the number of bits that have been set.

2. DualQuatSkinDeformer::Reinitialize and SoftSkinDeformer::ReInitialize were both iterating over a vector of joints each time a new jointId was added in order to determine if a joint with the same original Id exists already or if a new joint should be added. Instead, we pass in the value of the highest jointId, and construct a vector that allows us to directly look up whether or not a joint has already been added.

This results in a noticeable speedup during level load. Testing a level with the pyro asset,
Change 1 improved CalcLocalJointIndices from 334 samples (15.09%) to 23 samples (2.13%) for CalcLocalJointIndexCount
Change 2 improved Reinitialize from 438 samples (19.78%) to 22 samples (2.04%)

Profiling Results (sampling captured during level load with the pyro character):
Before
CalcLocalJoint_before

After
CalcLocalJoint_plusDualQuatReinit_after

### CODE REVIEW

File: Gems/EMotionFX/Code/EMotionFX/Source/DualQuatSkinDeformer.cpp [hunk 1]
```diff
@@ -335,21 +323,24 @@ namespace EMotionFX
             for (size_t a = 0; a < numInfluences; ++a)
             {
                 SkinInfluence* influence = skinningLayer->GetInfluence(i, a);
+                uint16 nodeIndex = influence->GetNodeNr();
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/EMotionFX/Source/DualQuatSkinDeformer.cpp:326]: nit: const for nodeindex and boneindex

File: Gems/EMotionFX/Code/EMotionFX/Source/DualQuatSkinDeformer.cpp [hunk 2]
```diff
@@ -324,6 +310,8 @@ namespace EMotionFX
         SkinningInfoVertexAttributeLayer* skinningLayer = (SkinningInfoVertexAttributeLayer*)m_mesh->FindSharedVertexAttributeLayer(SkinningInfoVertexAttributeLayer::TYPE_ID);
         MCORE_ASSERT(skinningLayer);
 
+        AZStd::vector<uint16> localBoneMap(highestJointId + 1, AZStd::numeric_limits<uint16>::max());
```
Comment (amzn-rhhong) [inline Gems/EMotionFX/Code/EMotionFX/Source/DualQuatSkinDeformer.cpp:313]: ultra nit: I would do a const uint16 invalidBoneIndex = AZStd::numeric_limits<uint16>::max() before line 313, then compare boneIndex to invalidBoneIndex in line 331.
Define a invalid bone index somewhere else would be great too.

File: Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformer.h [hunk 3]
```diff
@@ -49,7 +49,7 @@ namespace EMotionFX
          * @param node The node where the mesh belongs to during this initialization.
          * @param lodLevel The LOD level of the mesh the mesh deformer works on.
          */
-        virtual void Reinitialize(Actor* actor, Node* node, size_t lodLevel);
+        virtual void Reinitialize(Actor* actor, Node* node, size_t lodLevel, uint16 highestJointId);
```
[the remark this answers is not in the document]
Comment (amzn-jillich) [reply Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformer.h:53]: Can we call this index? Id feels confusing as we don't use that anywhere else and CalcLocalJointIndexCountAndHighestIndex() also says index.

File: Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformerStack.cpp [hunk 4]
```diff
@@ -110,10 +110,15 @@ namespace EMotionFX
...
+        EMotionFX::SkinningInfoVertexAttributeLayer* skinLayer =
+            (SkinningInfoVertexAttributeLayer*)m_mesh->FindSharedVertexAttributeLayer(SkinningInfoVertexAttributeLayer::TYPE_ID);
+        const AZ::u32 numOrgVerts = skinLayer->GetNumAttributes();
+        const auto [numLocalJoints, highestJointId] = skinLayer->CalcLocalJointIndexCountAndHighestIndex(numOrgVerts);
+
         // iterate through the deformers and reinitialize them
         for (size_t i = 0; i < numDeformers; ++i)
         {
-            m_deformers[i]->Reinitialize(actor, node, lodLevel);
+            m_deformers[i]->Reinitialize(actor, node, lodLevel, highestJointId);
```
Comment (amzn-jillich) [inline Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformerStack.cpp:121]: What about just initializing the local joint array with the number of joints in the skeleton and calculate the numLocalJoints while filling it, resizing/shrinking it afterwards?

That way we don't have to iterate through the skin influences twice, inside CalcLocalJointIndexCountAndHighestIndex as well as Reinitialize()

**Review by amzn-jillich — APPROVED** [review]
Great find!

File: Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformerStack.cpp (hunk 4 above)
Comment (amzn-tommy) [reply Gems/EMotionFX/Code/EMotionFX/Source/MeshDeformerStack.cpp]: I've updated this to calculate it while building the mesh and just retrieve it using a Get function later.

**Review by burelc-amzn — APPROVED** [review]

File: Gems/EMotionFX/Code/EMotionFX/Source/SkinningInfoVertexAttributeLayer.cpp [hunk 5]
```diff
@@ -329,4 +329,27 @@ namespace EMotionFX
 
         return result;
     }
+
+    AZStd::tuple<uint16, uint16> SkinningInfoVertexAttributeLayer::CalcLocalJointIndexCountAndHighestIndex(AZ::u32 numOrgVertices)
+    {
+        AZStd::bitset<AZStd::numeric_limits<uint16>::max()> joints;
+        uint16 highestJoint = 0;
+        for (AZ::u32 i = 0; i < numOrgVertices; i++)
```
[the remark this answers is not in the document]
Comment (burelc-amzn) [reply Gems/EMotionFX/Code/EMotionFX/Source/SkinningInfoVertexAttributeLayer.cpp]: Whoa that's weird, I dunno if these comments were still pending from a while ago, but you don't need to address these.

**Pull request merged**